File: observer-pattern/observer_pattern.rb
# frozen_string_literal: true

# When to use the Observer Pattern
#
# - when you need many other object to receive an update when
#   another object changes
# Ex:
#   - Stock market with thousands of stocks need to send updates to
#     objects representing individual stocks
#   - The Subject(publisher) sends many stocks to the observers
#   - The Observers(subscribers) takes the ones they want and use them
#
# @see http://www.newthinktank.com/2012/08/observer-design-pattern-tutorial/

class StockGrabber
  @observers = []

  attr_reader :observers, :stocks

  def initialize
    @observers = []
    @stocks = {}
  end

  # Subject Interface

  def register(observer)
    observers << observer
  end

  def unregister(observer)
    observers.delete(observer)
  end

  def notify_observers
    observers.each do |observer|
      observer.update(stocks)
    end
  end

  # StockGrabber specific methods

  def add_stock(name, price)
    stocks[name] = price unless stocks.key? name
    notify_observers
  end

  def remove_stock(name)
    stocks.delete(name)
    notify_observers
  end

  def update_stock(name, price)
    stocks[name] = price
    notify_observers
  end
end

class StockObserver
  @@observer_id_tracker = 0

  attr_reader :stocks

  def initialize(stock_grabber, stocks = {})
    @stocks = stocks
    @stock_grabber = stock_grabber
    @observer_id = @@observer_id_tracker += 1
    @stock_grabber.register(self)
  end

  def update(stocks)
    @stocks = stocks
    puts self
  end

  def to_s
    stock_prices_s = stocks.map { |name, price| "#{name}: #{price}" }.join("\n")
    <<~STRING
      #{'#' * 10} Observer: #{@observer_id} #{'#' * 10}
      #{stock_prices_s}
      #{'#' * 33}
    STRING
  end
end

################################################################################

stock_grabber = StockGrabber.new
observer_one = StockObserver.new(stock_grabber)

stock_grabber.add_stock('IBM', 10)
stock_grabber.add_stock('AAPL', 10)

_ = StockObserver.new(stock_grabber)

stock_grabber.add_stock('GOOG', 10)
stock_grabber.update_stock('GOOG', 20)

stock_grabber.unregister(observer_one)

stock_grabber.remove_stock('AAPL')

File: observer-pattern/observer-pattern.ts
/**
 * When to use the Observer Pattern
 *
 * - when you need many other object to receive an update when
 *   another object changes
 * Ex:
 *   - Stock market with thousands of stocks need to send updates to
 *     objects representing individual stocks
 *   - The Subject(publisher) sends many stocks to the observers
 *   - The Observers(subscribers) takes the ones they want and use them
 *
 * @see http://www.newthinktank.com/2012/08/observer-design-pattern-tutorial/
 */

interface Subject {
  register(o: Observer): void;
  unregister(o: Observer): void;
  notifyObservers(): void;
}

interface Observer {
  update(stocks: Object): void;
}

class DuplicateStockError extends Error {
  constructor(name: string) {
    super(`stock '${name}' already exists`);
  }
}

class StockNotFoundError extends Error {
  constructor(name: string) {
    super(`stock '${name}' doens't exists`);
  }
}

class StockGrabber implements Subject {
  private observers = new Array<Observer>();
  private stocks: Record<string, number> = {};

  register(o: Observer): void {
    this.observers.push(o);
  }

  unregister(o: Observer): void {
    const index = this.observers.indexOf(o);
    this.observers.splice(index, 1);
  }

  notifyObservers(): void {
    const stocksCopy = JSON.parse(JSON.stringify(this.stocks));
    this.observers.forEach((o) => o.update(stocksCopy));
  }

  addStock(name: string, price: number) {
    if (this.hasStock(name)) throw new DuplicateStockError(name);

    this.stocks[name] = price;
    this.notifyObservers();
  }

  removeStock(name: string) {
    if (!this.hasStock(name)) throw new StockNotFoundError(name);

    delete this.stocks[name];
    this.notifyObservers();
  }

  updateStock(name: string, price: number) {
    if (!this.hasStock(name)) throw new StockNotFoundError(name);

    this.stocks[name] = price;
    this.notifyObservers();
  }

  private hasStock(name: string) {
    return name in this.stocks;
  }
}

class StockObserver implements Observer {
  private stocks = new Object();

  constructor(grabber: StockGrabber) {
    grabber.register(this);
  }

  update(stocks: Object): void {
    console.log("got an update", { from: this.stocks, to: stocks });
    this.stocks = stocks;
  }
}

const grabber: StockGrabber = new StockGrabber();
const ob1: StockObserver = new StockObserver(grabber);
const ob2: StockObserver = new StockObserver(grabber);

grabber.addStock("foo", 123);
grabber.addStock("bar", 123);
grabber.addStock("baz", 123);
